Reading data spilled to disk may fail when multiple queries are ran in quick succession

noticed here https://github.com/pola-rs/polars/actions/runs/7446264538/job/20256144205?pr=13524
```
__________________________ test_out_of_core_sort_9503 __________________________
[gw0] linux -- Python 3.11.7 /home/runner/work/polars/polars/py-polars/.venv/bin/python
tests/unit/streaming/test_streaming_sort.py:137: in test_out_of_core_sort_9503
    df = q.collect(streaming=True)
polars/lazyframe/frame.py:1729: in collect
    return wrap_df(ldf.collect())
E   pyo3_runtime.PanicException: called `Option::unwrap()` on a `None` value
----------------------------- Captured stderr call -----------------------------
OOC sort forced
thread '<unnamed>' panicked at /home/runner/work/polars/polars/crates/polars-core/src/utils/mod.rs:559:34:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I can't reproduce the failure locally though

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -81,6 +81,7 @@ url = "2.4"
 version_check = "0.9.4"
 xxhash-rust = { version = "0.8.6", features = ["xxh3"] }
 zstd = "0.13"
+uuid = { version = "1.7.0", features = ["v4"] }
 
 polars = { version = "0.37.0", path = "crates/polars", default-features = false }
 polars-compute = { version = "0.37.0", path = "crates/polars-compute", default-features = false }

--- a/crates/polars-arrow/Cargo.toml
+++ b/crates/polars-arrow/Cargo.toml
@@ -78,7 +78,6 @@ arrow-data = { workspace = true, optional = true }
 arrow-schema = { workspace = true, optional = true }
 
 [dev-dependencies]
-apache-avro = { version = "0.16", features = ["snappy"] }
 criterion = "0.5"
 crossbeam-channel = { workspace = true }
 doc-comment = "0.3"

--- a/crates/polars-core/src/utils/mod.rs
+++ b/crates/polars-core/src/utils/mod.rs
@@ -561,6 +561,21 @@ pub fn get_time_units(tu_l: &TimeUnit, tu_r: &TimeUnit) -> TimeUnit {
     }
 }
 
+pub fn accumulate_dataframes_vertical_unchecked_optional<I>(dfs: I) -> Option<DataFrame>
+where
+    I: IntoIterator<Item = DataFrame>,
+{
+    let mut iter = dfs.into_iter();
+    let additional = iter.size_hint().0;
+    let mut acc_df = iter.next()?;
+    acc_df.reserve_chunks(additional);
+
+    for df in iter {
+        acc_df.vstack_mut_unchecked(&df);
+    }
+    Some(acc_df)
+}
+
 /// This takes ownership of the DataFrame so that drop is called earlier.
 /// Does not check if schema is correct
 pub fn accumulate_dataframes_vertical_unchecked<I>(dfs: I) -> DataFrame

--- a/crates/polars-pipe/Cargo.toml
+++ b/crates/polars-pipe/Cargo.toml
@@ -19,6 +19,7 @@ polars-plan = { workspace = true }
 polars-row = { workspace = true }
 polars-utils = { workspace = true, features = ["sysinfo"] }
 tokio = { workspace = true, optional = true }
+uuid = { workspace = true }
 
 crossbeam-channel = { workspace = true }
 crossbeam-queue = { workspace = true }

--- a/crates/polars-pipe/src/executors/sinks/io.rs
+++ b/crates/polars-pipe/src/executors/sinks/io.rs
@@ -91,7 +91,9 @@ fn gc_thread(operation_name: &'static str) {
                         eprintln!("could not modified time on this platform")
                     }
                 } else {
-                    std::fs::remove_dir_all(path).unwrap()
+                    // This can be fallible as another Polars query could already have removed this.
+                    // So we ignore the result.
+                    let _ = std::fs::remove_dir_all(path);
                 }
             }
         }

--- a/crates/polars-pipe/src/executors/sinks/io.rs
+++ b/crates/polars-pipe/src/executors/sinks/io.rs
@@ -37,13 +37,10 @@ fn get_lockfile_path(dir: &Path) -> PathBuf {
 }
 
 fn get_spill_dir(operation_name: &'static str) -> PolarsResult<PathBuf> {
-    let uuid = SystemTime::now()
-        .duration_since(SystemTime::UNIX_EPOCH)
-        .unwrap()
-        .as_nanos();
+    let id = uuid::Uuid::new_v4();
 
     let mut dir = std::path::PathBuf::from(get_base_temp_dir());
-    dir.push(&format!("polars/{operation_name}/{uuid}"));
+    dir.push(&format!("polars/{operation_name}/{id}"));
 
     if !dir.exists() {
         fs::create_dir_all(&dir).map_err(|err| {

--- a/crates/polars-pipe/src/executors/sinks/sort/ooc.rs
+++ b/crates/polars-pipe/src/executors/sinks/sort/ooc.rs
@@ -4,7 +4,9 @@ use std::sync::atomic::{AtomicU32, AtomicU64, Ordering};
 use crossbeam_queue::SegQueue;
 use polars_core::prelude::*;
 use polars_core::series::IsSorted;
-use polars_core::utils::accumulate_dataframes_vertical_unchecked;
+use polars_core::utils::{
+    accumulate_dataframes_vertical_unchecked, accumulate_dataframes_vertical_unchecked_optional,
+};
 use polars_core::POOL;
 use polars_io::ipc::IpcReader;
 use polars_io::SerReader;
@@ -54,7 +56,8 @@ impl PartitionSpillBuf {
             // so we pop no more than the current size.
             let pop_max = len;
             let iter = (0..pop_max).flat_map(|_| self.chunks.pop());
-            Some(accumulate_dataframes_vertical_unchecked(iter))
+            // Due to race conditions, the chunks can already be popped, so we use optional.
+            accumulate_dataframes_vertical_unchecked_optional(iter)
         } else {
             None
         }

